### ISSUE TITLE
Switching animation direction is wrong in Gnome 3.30

### ISSUE THREAD

**Issue by thijzert** [zakkak/workspace-grid#88]
When changing workspaces in the X direction (e.g. to the right), the windows move as if you're switching to a workspace in the Y direction (e.g. down).

I've recorded a screencast showcasing the problem:
![Screencast](https://user-images.githubusercontent.com/845572/45809006-7313ee80-bcc7-11e8-9dee-26b25cb7a024.gif)

**Comment by thijzert** [comment]
(Note, the extension is currently broken in 3.30, see #86. For now you can get the extension working by checking out the branch referenced [here](https://github.com/zakkak/workspace-grid/issues/86#issuecomment-423074499) and following the instructions in [this comment](https://github.com/zakkak/workspace-grid/issues/86#issuecomment-423028882).)

**Comment by renxida** [comment]
[ just in case you have Gnome 2.28 and are experiencing the same problems]

If you installed the extension from the gnome extension website https://extensions.gnome.org/extension/484/workspace-grid/, you might have an old version.

Clone the newest commit on branch 2.28, cd into it, and do `make zip`

Then proceed to unzip the contents to `/home/xren/.local/share/gnome-shell/extensions`. Don't forget to remove the old version beforehand.

The newest version in 2.28 has all the animations right.

**Comment by rmyorston** [comment]
There's a problem with workspace layouts in mutter 3.30:

[https://gitlab.gnome.org/GNOME/mutter/issues/270](https://gitlab.gnome.org/GNOME/mutter/issues/270)

**Comment by jonnixs** [comment]
The [MR](https://gitlab.gnome.org/GNOME/mutter/merge_requests/245) for that mutter issue has been merged. Does that allow this bug to get resolved?

**Comment by thijzert** [comment]
I manually applied the patch in mutter to my system, and it's working great! Closing the issue here, since there's nothing to be done in this extension; the patch will eventually make its way downstream.

**Comment by mzur** [comment]
@thijzert I tried to build Mutter from source to see if the patch solves a [similar issue](https://github.com/mzur/gnome-shell-wsmatrix/issues/2#issuecomment-434449606) for me. I could not find any installation instructions anywhere so I had to find out all the dependencies the hard way. Now I'm stuck with `No rule to make target 'Clutter-3.typelib', needed by 'all-am'.` and don't know how to continue. Can you please give me some hints on how you manage to build Mutter on your system?

**Comment by thijzert** [comment]
I'm on Arch Linux, so for me it was a matter of getting the `PKGBUILD` from the Arch Build System, modifying it to apply the patch, and have `makepkg` take care of the rest. For other distributions, I'm not sure; I think you're better off just waiting for the Gnome team to tag the next release and for it to percolate to your OS vendor. (It was merged to master just two days ago, mind you.)

**Comment by mzur** [comment]
I was hoping to confirm that the patch fixes the issue but setting up my own Arch VM just for this is a little too much. I guess I'll do as you say and wait until the patch arrives downstream. Thanks for your help!

**Comment by triffon** [comment]
I am using Ubuntu 18.10 and can confirm that applying the patch against the repository version of mutter solves the problem for me. Thank you!

**Comment by LorincJuraj** [comment]
@triffon can you give us a walk-through how this can be done? I am on Ubuntu 18.10 as well would prefer not have to wait for official update.

**Comment by triffon** [comment]
Sure, please see below:

0. Install the Debian build tools: `sudo apt install devscripts`
1. Download the package source: `apt source mutter`
2. Install build dependencies: `sudo apt build-dep mutter`
3. Navigate to the newly created directory: `cd mutter-3.30.1-1`
4. The patch is only one line, so you can edit the file directly: `gedit src/core/meta-workspace-manager.c`
5. Add the following line in the beginning of the function `meta_workspace_manager_override_workspace_layout` (line 499): `workspace_manager->workspace_layout_overridden = FALSE;`
6. Build the package: `dpkg-buildpackage -us -uc -nc`
6. (Optional) Add a new changelog entry: `dch`. The actual description is not important, what matters is that a new changelog entry with a new version is created.
7. Install the new packages: `sudo dpkg -i ../*.deb`
8. Restart and enjoy!

**Comment by mzur** [comment]
Brilliant, thank you! @LorincJuraj Note that you need to [check "source code"](https://askubuntu.com/questions/496549/error-you-must-put-some-source-uris-in-your-sources-list) in Software & Updates for `apt source` to work.

**Comment by LorincJuraj** [comment]
Thanks @triffon and @mzur . It worked. The only difference was in step 3 I had to go to `cd mutter-3.30.1` instead. Step 7 didn't work for me but as it was optional, I hope it will be OK.

**Comment by akyoto** [comment]
@triffon Worked flawlessly (step 3 replaced with `3.30.1`), thank you!

**Comment by amcsi** [comment]
@triffon I tried following your steps, but I believe I'm stuck at step 2:

```
  ~ git:(master) ✗ apt source mutter
Reading package lists... Done
E: You must put some 'source' URIs in your sources.list
➜  ~ git:(master) ✗ sudo apt build-dep mutter
Reading package lists... Done
E: You must put some 'source' URIs in your sources.list
```

**Comment by mzur** [comment]
@amcsi Check out my [comment](https://github.com/zakkak/workspace-grid/issues/88#issuecomment-435073547) above.

**Comment by harish2704** [comment]
I reported this bug in mutter https://gitlab.gnome.org/GNOME/mutter/issues/409

**Comment by tstirrat15** [comment]
This seemed like it was fixed for a while, and then it just stopped working for me again today. Was there another change to Mutter recently?

**Comment by tstirrat15** [comment]
I went through @triffon's steps again, and it fixed it again. The difference is that rather than adding the `stuff = FALSE` statement, you're changing the `stuff = TRUE` statement to `FALSE`. Any idea why they broke it again?

**Comment by LorincJuraj** [comment]
Again only small change in step 3 use `cd mutter-3.30.2`

**Comment by akyoto** [comment]
Seems to be bugged again there.
I don't know if it was ever fixed, but the [comment](https://gitlab.gnome.org/GNOME/mutter/issues/409) said it was.

**Comment by NeroVanbiervliet** [comment]
Check your `mutter --version`. My distro is still on `3.30.2`. According to gitlab, the fix is merged starting from  `3.31.2`.

**Comment by kevinmarrec** [comment]
Hi guys, What's the current fix then ?

Fix Steps seem to have been changing multiple times since @triffon last comment, and the fix might also not being available anymore (assuming @tstirrat15 feedback)  ?

**Comment by LorincJuraj** [comment]
@kevinmarrec Triffon's solution worked for me last time (see my last comment).

**Comment by kevinmarrec** [comment]
@LorincJuraj Alright, thanks for blazing fast answer !

**Comment by tstirrat15** [comment]
Most recent `mutter` update in Ubuntu 18.10 broke things again. Is this being tracked somewhere on the `mutter` project, or is this an issue with Ubuntu being behind on releases?

**Comment by kevinmarrec** [comment]
@tstirrat15  Ubuntu 18.10 has indeed upgraded `mutter` from `3.30.2-1~ubuntu18.10.3` to `mutter_3.30.2-1~ubuntu18.10.4`lately.
The fix is only in `mutter@3.31.2` and above, so we still need to manually to apply the `fix` ourselves (`workspace_layout_overridden = FALSE`), rebuild source and reinstall each time they'll update Ubuntu 18.10.x

(FYI Ubuntu 19.04 should be released in early April with expected Gnome 3.32 so no more issues)

**Comment by jtagcat** [comment]
> Sure, please see below:
> 
>     1. Install the Debian build tools: `sudo apt install devscripts`
> 
>     2. Download the package source: `apt source mutter`
> 
>     3. Install build dependencies: `sudo apt build-dep mutter`
> 
>     4. Navigate to the newly created directory: `cd mutter-3.30.1-1`
> 
>     5. The patch is only one line, so you can edit the file directly: `gedit src/core/meta-workspace-manager.c`
> 
>     6. Add the following line in the beginning of the function `meta_workspace_manager_override_workspace_layout` (line 499): `workspace_manager->workspace_layout_overridden = FALSE;`
> 
>     7. Build the package: `dpkg-buildpackage -us -uc -nc`
> 
>     8. (Optional) Add a new changelog entry: `dch`. The actual description is not important, what matters is that a new changelog entry with a new version is created.
> 
>     9. Install the new packages: `sudo dpkg -i ../*.deb`
> 
>     10. Restart and enjoy!

How would you do this on Fedora?
